File: 1122-relative-sort-array/1122-relative-sort-array.java
class Solution {
    public int[] relativeSortArray(int[] arr1, int[] arr2) {
        Map<Integer, Integer> orderMap = new HashMap<>();
        int N = arr1.length;
        int M = arr2.length;
        for(int i = 0; i < M; i++) {
            orderMap.put(arr2[i], i);
        }
        //System.out.println(orderMap);
        Arrays.sort(arr1);
        int tmp; boolean flag;
        for(int re = 0; re < N; re++) {
            flag = false;
            for(int i = 0; i < N - 1; i++) {
                if(!orderMap.containsKey(arr1[i])) {
                    orderMap.put(arr1[i], M++);
                    //System.out.println(orderMap);
                }
                if(!orderMap.containsKey(arr1[i + 1])) {
                    orderMap.put(arr1[i + 1], M++);
                    //System.out.println(orderMap);
                }
                if(orderMap.get(arr1[i]) > orderMap.get(arr1[i + 1])) {
                    flag = true;
                    tmp = arr1[i];
                    arr1[i] = arr1[i + 1];
                    arr1[i + 1] = tmp;
                }
            }
            if(!flag) break;
        }
        
        return arr1;
    }
}

File: 1122-relative-sort-array/1122-relative-sort-array.cpp
class Solution {
private:
    map<int, int> orderMap;
public:
    vector<int> relativeSortArray(vector<int>& arr1, vector<int>& arr2) {
        int N = arr1.size();
        int M = arr2.size();
        for(int i = 0; i < M; i++) {
            orderMap.emplace(arr2[i], i);
        }
        
        sort(arr1.begin(), arr1.end());
        for(int i = 0; i < N; i++) {
            if(orderMap.find(arr1[i]) == orderMap.end()) {
                orderMap.emplace(arr1[i], M++);
            }
        }
        
        sort(arr1.begin(), arr1.end(), [this](int a, int b) {
            return orderMap[a] < orderMap[b];
        });
        return arr1;
    }
};